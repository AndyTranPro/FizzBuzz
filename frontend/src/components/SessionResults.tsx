import React, { useState, useEffect } from 'react';
import { useAppSelector } from '../features/store';
import { useNavigate } from 'react-router-dom';
import { Typography, Button, Box, Paper } from '@mui/material';
import CheckCircleOutlineIcon from '@mui/icons-material/CheckCircleOutline';
import SentimentDissatisfiedIcon from '@mui/icons-material/SentimentDissatisfied';
import Lottie from 'react-lottie';
import animationData from '../animations/congrats.json';


// A component to display the results of the session

const SessionResults: React.FC = () => {
    const score = useAppSelector((state) => state.session.score);
    const navigate = useNavigate();
    const totalQuestions = useAppSelector((state) => state.session.numQuestions);
    const [message, setMessage] = useState('');
    const [isCongrats, setIsCongrats] = useState(false);

    useEffect(() => {
        const percentage = (score / totalQuestions) * 100;
        if (percentage >= 80) {
            setMessage('Congratulations! You did a great job!');
            setIsCongrats(true);
        } else if (percentage < 50) {
            setMessage('Better luck next time!');
            setIsCongrats(false);
        }
    }, [score, totalQuestions]);

    const BackToHomePage = () => {
        navigate("/");
    };

    const defaultOptions = {
        loop: true,
        autoplay: true,
        animationData: animationData,
        rendererSettings: {
            preserveAspectRatio: 'xMidYMid slice'
        }
    };

    return (
        <Box
            sx={{
                display: 'flex',
                flexDirection: 'column',
                alignItems: 'center',
                justifyContent: 'center',
                height: '80vh',
                textAlign: 'center',
            }}
        >
            <Paper elevation={5} sx={{ p: 4, maxWidth: 600, width: '100%' }}>
                <Typography
                    variant="h4"
                    gutterBottom
                    sx={{
                        background: 'linear-gradient(45deg, #FE6B8B 30%, #FF8E53 90%)',
                        WebkitBackgroundClip: 'text',
                        WebkitTextFillColor: 'transparent',
                        fontFamily: 'Roboto, sans-serif',
                        fontWeight: 'bold',
                    }}
                >
                    Game Over!
                </Typography>
                {isCongrats ? (
                    <>
                        <Lottie options={defaultOptions} height={150} width={150} />
                        <Typography variant="h6" component="h2" sx={{ mt: 2 }}>
                            {message}
                        </Typography>
                        <CheckCircleOutlineIcon sx={{ fontSize: 50, color: 'green', mt: 2 }} />
                    </>
                ) : (
                    <>
                        <SentimentDissatisfiedIcon sx={{ fontSize: 50, color: 'red', mt: 2 }} />
                        <Typography variant="h6" component="h2" sx={{ mt: 2 }}>
                            {message}
                        </Typography>
                    </>
                )}
                <Typography variant="h6" gutterBottom>
                    Your Score: {score}/{totalQuestions}
                </Typography>
                <Button
                    onClick={BackToHomePage}
                    variant="contained"
                    color="primary"
                    sx={{
                        mt: 2,
                        background: 'linear-gradient(45deg, #2196F3 30%, #21CBF3 90%)',
                        color: 'white',
                        fontWeight: 'bold',
                        '&:hover': {
                            background: 'linear-gradient(45deg, #21CBF3 30%, #2196F3 90%)',
                        },
                    }}
                >
                    Back to Home
                </Button>
            </Paper>
        </Box>
    );
};

export default SessionResults;