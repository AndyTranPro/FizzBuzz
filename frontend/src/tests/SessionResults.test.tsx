import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import SessionResults from '../components/SessionResults';

jest.mock('react-lottie', () => () => <div>Lottie Animation</div>);

const mockStore = configureStore([]);

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));

const renderComponent = (store: any) => {
    return render(
        <Provider store={store}>
            <MemoryRouter>
                <SessionResults />
            </MemoryRouter>
        </Provider>
    );
};

describe('SessionResults Component', () => {
    let store: any;

    beforeEach(() => {
        store = mockStore({
            session: {
                score: 8,
                numQuestions: 10,
            },
        });
    });

    test('renders correctly and matches snapshot', () => {
        const { asFragment } = renderComponent(store);
        expect(asFragment()).toMatchSnapshot();
    });

    test('displays congratulatory message when score is 80% or higher', () => {
        renderComponent(store);

        expect(screen.getByText(/Congratulations! You did a great job!/i)).toBeInTheDocument();
        expect(screen.getByText(/Your Score: 8\/10/i)).toBeInTheDocument();
        expect(screen.getByText(/Lottie Animation/i)).toBeInTheDocument();
        expect(screen.getByTestId('CheckCircleOutlineIcon')).toBeInTheDocument();
    });

    test('displays condolence message when score is less than 50%', () => {
        store = mockStore({
            session: {
                score: 4,
                numQuestions: 10,
            },
        });
        renderComponent(store);

        expect(screen.getByText(/Better luck next time!/i)).toBeInTheDocument();
        expect(screen.getByText(/Your Score: 4\/10/i)).toBeInTheDocument();
        expect(screen.getByTestId('SentimentDissatisfiedIcon')).toBeInTheDocument();
    });

    test('back to home page', () => {

        renderComponent(store);

        const button = screen.getByRole('button', { name: /Back to Home/i });
        fireEvent.click(button);

        expect(mockNavigate).toHaveBeenCalledWith('/');
    });
});
