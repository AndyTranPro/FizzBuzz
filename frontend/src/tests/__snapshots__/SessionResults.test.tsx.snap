// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionResults Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1xei2fu"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 css-1reax9b-MuiPaper-root"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-dk1kvi-MuiTypography-root"
      >
        Game Over!
      </h4>
      <div>
        Lottie Animation
      </div>
      <h2
        class="MuiTypography-root MuiTypography-h6 css-1ry679h-MuiTypography-root"
      >
        Congratulations! You did a great job!
      </h2>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-gzwv8z-MuiSvgIcon-root"
        data-testid="CheckCircleOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 7.58 10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8"
        />
      </svg>
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
      >
        Your Score: 8/10
      </h6>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1u6ee9p-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Back to Home
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
